src/autoschedulers/anderson2021/generate_data.sh

```diff
@@ -0,0 +1,230 @@
+# It is a wrapper around autotune_loop.sh, which handles compiling, benchmarking,
+# and retraining the cost model. This file makes the process more user friendly
+# by providing statistics, support for resuming previous batches, autotuning
+# across multiple apps, etc.
+#
+# It assumes that the autoscheduler itself and any apps to be autoscheduled have
+# already been built and the resulting files are stored in halide_build_dir.
+# Using CMake is recommended because this script assumes that the given
+# halide_build_dir has the same structure that the CMake build will produce
+#
+# Arguments:
+# halide_build_dir - path where Halide is built
+# max_iterations - the number of batches to generate. The cost model is
+# retrained after each
+# resume - resume using the previously generated samples or start a new run?
+# train_only - don't generate new data, just retrain the cost model with
+# existing samples
+# predict_only - don't generate new data, just predict the costs of the existing
+# samples
+# parallelism - the number of streaming multiprocessors in the target GPU
+# app - the individual application (in Halide/apps/) to generate data for. If
+# not provided, it will generate a data for all the apps in the list below
+
+if [[ $# -ne 6 && $# -ne 7 ]]; then
+    echo "Usage: $0 halide_build_dir max_iterations resume train_only predict_only parallelism app"
+    exit
+fi
+
+set -eu
+
+HALIDE_BUILD_DIR=${1}
+MAX_ITERATIONS=${2}
+RESUME=${3}
+TRAIN_ONLY=${4}
+PREDICT_ONLY=${5}
+PARALLELISM=${6}
+
+if [ -z ${7+x} ]; then
+    APPS="bgu bilateral_grid local_laplacian nl_means lens_blur camera_pipe stencil_chain harris hist max_filter unsharp interpolate conv_layer cuda_mat_mul iir_blur depthwise_separable_conv"
+else
+    APPS=${7}
+fi
+
+if [ -z ${CXX+x} ]; then
+    echo The CXX environment variable must be set. Exiting...
+    exit
+fi
+
+export CXX="ccache ${CXX}"
+
+AUTOSCHEDULER_SRC_DIR=$(dirname $0)
+SCRIPTS_DIR="${AUTOSCHEDULER_SRC_DIR}/scripts"
+source ${SCRIPTS_DIR}/utils.sh
+make_dir_path_absolute $(dirname $0) AUTOSCHEDULER_SRC_DIR
+
+get_halide_src_dir ${AUTOSCHEDULER_SRC_DIR} HALIDE_SRC_DIR
+get_autoscheduler_build_dir ${HALIDE_BUILD_DIR} AUTOSCHEDULER_BUILD_DIR
+
+echo "HALIDE_SRC_DIR = ${HALIDE_SRC_DIR}"
+echo "HALIDE_BUILD_DIR = ${HALIDE_BUILD_DIR}"
+echo "AUTOSCHEDULER_SRC_DIR = ${AUTOSCHEDULER_SRC_DIR}"
+echo "AUTOSCHEDULER_BUILD_DIR = ${AUTOSCHEDULER_BUILD_DIR}"
+echo
+
+BEST_SCHEDULES_DIR=${AUTOSCHEDULER_SRC_DIR}/best
+export HL_PERMIT_FAILED_UNROLL=1
+
+if [ -z ${HL_TARGET+x} ]; then
+    get_host_target ${AUTOSCHEDULER_BUILD_DIR} HL_TARGET
+    HL_TARGET=${HL_TARGET}-cuda-cuda_capability_70
+fi
+
+echo "HL_TARGET set to ${HL_TARGET}"
+echo
+
+if [[ $PREDICT_ONLY == 1 && $TRAIN_ONLY == 1 ]]; then
+    echo "At most one of train_only and predict_only can be set to 1."
+    exit
+fi
+
+if [[ $PREDICT_ONLY == 1 ]]; then
+    echo "Predict only mode: ON"
+fi
+
+DEFAULT_SAMPLES_DIR_NAME="${SAMPLES_DIR:-autotuned_samples}"
```

Comment:
This will wail if the user has called this script with `train_only == 1` as the executable for retraining the cost model will silently fail loading empty samples and generate a hard to parse error message.

Reply:
hard-to-parse error messages are not good, please fix :-)